scripts/test-e2e.mjs

```diff
@@ -0,0 +1,153 @@
+  const { status } = spawnSync(command, args, { stdio: "inherit" });
+  if (status !== 0) {
+    throw new Error(
+      `An error occurred while executing: ${command} ${args.join(" ")}`
+    );
+  }
+}
+
+/**
+ * Invokes a shell command with optional arguments. Similar {@link $}, but
+ * captures and returns stdout/stderr.
+ * @param {string} command
+ * @param {...string} args
+ * @returns {string}
+ */
+function $$(command, ...args) {
+  const { status, stderr, stdout } = spawnSync(command, args, {
+    stdio: ["ignore", "pipe", "pipe"],
+    encoding: "utf-8",
+  });
+  if (status !== 0) {
+    throw new Error(
+      `An error occurred while executing: ${command} ${args.join(" ")}`
+    );
+  }
+  // Some commands, like `appium driver list --installed` output to stderr
+  return stdout.trim() || stderr.trim();
+}
+
+/**
+ * Ensures Appium is available.
+ * @returns {Promise<void>}
+ */
+function ensureAppiumAvailable() {
+  return new Promise((resolve, reject) => {
+    const socket = new Socket();
+    socket.setTimeout(60);
+
+    const onError = () => {
+      socket.destroy();
+      reject(new Error("Could not find Appium server"));
+    };
+
+    socket.once("error", onError);
+    socket.once("timeout", onError);
+
+    socket.connect(4723, "localhost", () => {
+      socket.end();
+      resolve();
+    });
+  });
+}
+
+/**
+ * Install or update specified Appium driver.
+ */
+const installAppiumDriver = (() => {
+  const appium = new URL("../example/node_modules/.bin/appium", import.meta.url)
+    .pathname;
+
+  const installedDrivers = /** @type {string[]} */ ([]);
+  /** @type {(driver: string) => void} */
+  return (driver) => {
+    if (!installedDrivers.includes(driver)) {
+      installedDrivers.push(driver);
+      const drivers = $$(appium, "driver", "list", "--installed");
+      const action = drivers.includes(driver) ? "update" : "install";
+      $(appium, "driver", action, driver);
+    }
+  };
+})();
+
+function prepareAndroid(androidHome = process.env["ANDROID_HOME"]) {
+  // Note: Ubuntu agents can't run Android emulators — see
+  // https://github.com/actions/runner-images/issues/6253#issuecomment-1255952240
+  const adb = androidHome + "/platform-tools/adb";
+
+  /*
+  const android_image = "system-images;android-30;google_atd;x86_64";
+  const avdmanager = androidHome + "/cmdline-tools/latest/bin/avdmanager";
+  const emulator = androidHome + "/emulator/emulator";
+  const sdkmanager = androidHome + "/cmdline-tools/latest/bin/sdkmanager";
+
+  if [[ -n $CI ]]; then
+    # Accept all licenses so we can download an Android image
+    yes 2> /dev/null | $sdkmanager --licenses
+    $sdkmanager --install "$android_image"
+
+    # Create an Android emulator and boot it up
+    echo "no" | $avdmanager create avd --package "$android_image" --name Android_E2E
+    $emulator @Android_E2E -delay-adb -partition-size 4096 -no-snapshot -no-audio -no-boot-anim -no-window -gpu swiftshader_indirect &
+  fi
+   */
```

Comment:
do we need this still?

Reply:
Yeah, I wanted to keep it around in case we want to try getting this running on CI again. It was a pain to figure out so I'd rather not lose it.